Add first half of Getting Started for Golang

olliefr opened the pull request:
### Proposed changes

There's currently no "Getting Started" guide for Go programming language.

So I've written a first half of "Getting Started" guide for Go, using Node.js guide as a template. Before I spend more time to finish the guide, I'd like to find out if what I've already done is of acceptable quality and has a chance to be included in `master`. Actually, I don't even know if it's needed at all, but it was a fun thing to do 🙃 

Current progress:

- [x] Build images
- [x] Run your image as a container
- [x] Use containers for development (80% complete)
- [x] Run your tests
- [x] Configure CI/CD
- [x] Deploy your app

### Related issues (optional)

N/A, as far as I can tell.

usha-mandya commented:
@olliefr This is great, thanks. I'll take a look and get back to you soon.

StefanScherer reviewed (changes requested):
Hey @olliefr 
Thank you very much for your contribution! ❤️
That's a really great idea and we absolutely want it published in our docs site, so other Go developers can start their journey with Docker.
I left a few minor comments and suggestions, but all in all it already looks really good!
We really appreciate your work and time and of course would love to have the second part as well if you find some time.

usha-mandya commented:
Hey @olliefr Wondering if you got a chance to take a look at the suggestions above by Stefan. Please reach out to us if you have any questions. Look forward to getting the second part of this guide if you have time. Thanks again :)

olliefr commented on language/golang/build-images.md:
```diff
@@ -0,0 +1,354 @@
+
+To complete this tutorial, you need the following:
+
+- Go version 13 or later: [Download Go](https://golang.org/dl/){: target="_blank" rel="noopener" class="_"}.
```
Done)

olliefr commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,354 @@
+RUN go mod download
+```
+
+At this point, we have an image that is based on Go environment version 13.1 (or later minor version) and we have installed our dependencies. The next thing we need to do is to add our source code into the image. We’ll use the COPY command just like we did with our module files above.
````
Done)

olliefr commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,354 @@
+Open this working directory in your IDE and add the following code into the `main.go` file.
+
+```go
+// TODO update with GET /users and server status
````
Removed)

olliefr commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,354 @@
+  --url http://localhost:8000/users \
+  --header 'content-type: application/json' \
+  --data '{"name": "The Mouse", "email": "mouse@deepdarkwood" }'
+```
````
Not sure about this one. I've replaced the whole example, if the problem persists, I'd need to figure out how to dump more debug information into the output.

olliefr commented on language/golang/develop.md:
````diff
@@ -0,0 +1,250 @@
+Now we can build our image.
+
+```shell
+$ docker build --tag node-docker .
````
"Develop" part is under active development, I'll update & resolve this point when I come to it)

olliefr replied:
That part was redrafted heavily and no longer includes this code.

olliefr commented on language/golang/develop.md:
```diff
@@ -0,0 +1,250 @@
+  --name rest-server \
+  -p 8000:8000 \
+  -e CONNECTIONSTRING=mongodb://mongodb:27017/yoda_notes \
+  node-docker
```
"Develop" part is under active development, I'll update & resolve this point when I come to it)

olliefr replied:
That part was redrafted heavily and no longer includes this code.

olliefr commented on language/golang/develop.md:
```diff
@@ -0,0 +1,250 @@
+
+## Use Compose to develop locally
+
+In this section, we’ll create a Compose file to start our node-docker and the MongoDB with one command. We’ll also set up the Compose file to start the node-docker in debug mode so that we can connect a debugger to the running node process.
```
"Develop" part is under active development, I'll update & resolve this point when I come to it)

olliefr replied:
That part was redrafted heavily and no longer includes this text.

olliefr commented on language/golang/run-containers.md:
````diff
@@ -0,0 +1,187 @@
+Start the container and expose port 8000 to port 8000 on the host.
+
+```shell
+$ docker run --publish 8000:8000 node-docker
````
Fixed!

olliefr commented on language/golang/run-tests.md:
````diff
@@ -0,0 +1,259 @@
+Below is the Docker command to start the container and run tests:
+
+```shell
+$ docker run -it --rm --name app -p 8080:80 node-docker npm run test
````
"Run tests" part is under active development, I'll update & resolve this point when I come to it)

olliefr replied:
That part was redrafted heavily and no longer includes this code.

olliefr commented on language/golang/run-tests.md:
````diff
@@ -0,0 +1,259 @@
+Now that we have our sample application locally, let’s build our Docker image and make sure everything is running properly. Run the following commands to build and then run your Docker image in a container.
+
+```shell
+$ docker build -t node-docker .
````
"Run tests" part is under active development, I'll update & resolve this point when I come to it)

olliefr replied:
That part was redrafted heavily and no longer includes this code.

olliefr commented:
Hey @usha-mandya thank you for the reminder! I've fixed some of the points raised in the review and am working on the rest. The "Develop" page is the next one to be completed. I'm working on a (simple) example which connects a Go program running in a Docker container with a database (CockroachDB) running in another.

I've switched the PR to draft mode to avoid distracting the reviewers until the whole guide is complete. My original intention with the PR was to gauge the level of interest and now having established that contributions are welcome, I can finish the guide, hopefully over this coming weekend :shipit:

usha-mandya commented:
That sounds great. Thank you @olliefr

olliefr commented:
Just a progress update. I'll try to keep it brief 🙃 

I've added the bulk of the _Develop_ section, including an example of a Go application which provides a REST API endpoint using CockroachDB as a database. Naturally, the application and the database run in two different containers. There is also a Docker Compose configuration for this set-up. 

I'd like to have another go at this section, because it's a bit rough around the corners. But the example presented is realistic and the Docker Compose set up is similar to what I am using in real projects, especially in the way environment variables are used in conformance with the [12 Factor App](https://12factor.net/) philosophy.

I've also created a [repo for the code example](https://github.com/olliefr/go-docker-example). It has two branches. The `main` branch is for the first version of the application, which is built using the Dockerfile. The `roach` branch is more sophisticated and is meant to be run together with CockroachDB, and there is a Docker Compose configuration provided for that.

Other parts of the guide are coming up soon as well, time permitting. I'm really excited about this opportunity to contribute to Docker documentation. Thank you for your patience)

StefanScherer commented:
@olliefr Thank you very much! We're really excited about your contribution. 😍 
I'll take a closer look next week. I also want to test drive this on an Apple M1.

olliefr commented:
Hey 🙃 

My apologies for taking my time to progress on this. But I do have some updates.

I've dramatically reduced the complexity (and size) of one of the example applications, which prompted me to rewrite large chunks of previously completed sections. I think they are shorter and less confusing now.

This comes with new benefits though. Now I had the space to cover new material, which is going to be useful for Go developers:

* Using multi-stage builds to produce very lean images.
* Using `ory/dockertest` for functional testing.

The latter is a great start for coming improvements to the CI/CD section, because I can use those tests in GitHub Actions. This, combined with the multi-stage build example, would allow me to demonstrate the pipelines for testing with GitHub runners, as well as deploying lean static images to the Docker Hub. That would be a pretty flexible and a quite realistic CI/CD pipeline.

I still have to do some work, especially on the second example I am using, which uses a database.

The things that are left for me to do:

- [ ] Review the `develop.md` part; I'm pretty sure the CockroachDB example can be simplified and reduced in length.
- [ ] Write the `configure-ci-cd.md` part, using GitHub Actions in a manner described above.

I hope to finish before the end of this week. I'm really anxious to get this into prod ASAP. I'll keep you posted.

StefanScherer commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,388 @@
+CMD [ "/docker-gs-ping" ]
+```
+
+You can also find this `Dockerfile` for this part in [go-docker](https://github.com/olliefr/go-docker) repo.
````
The link is wrong, should it be https://github.com/olliefr/docker-gs-ping ?

olliefr replied:
That sentence shouldn't be there at all. Removed!

StefanScherer commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,388 @@
+For our present study, we clone it to our local machine:
+
+```shell
+$ git clone github.com/olliefr/docker-gs-ping
````
Add `https://` to make `git clone` work.

```suggestion
$ git clone https://github.com/olliefr/docker-gs-ping
```

olliefr replied:
Interesting point! Fixed.

StefanScherer commented on language/golang/build-images.md:
````diff
@@ -0,0 +1,388 @@
+Let's build our first Docker image!
+
+```shell
+$ docker build --tag docker-gs-ping .
````
Should we keep `go-docker` as local image name? It is used later to update the image.

olliefr replied:
`go-docker` was a different application, it's been retired since 😂 I'm sticking to the ping example for the first part and CI/CD. And I'll have a different example for the "Develop" part since it involves the database.

StefanScherer commented on language/golang/develop.md:
````diff
@@ -0,0 +1,475 @@
+We can build an updated image with the following command.
+
+```shell
+$ docker build --tag go-docker .
````
Use same image name `docker-gs-ping` as before?

olliefr replied:
It's a different application. I was trying to use the same application throughout, but introducing a database adds so much complexity, that I decided to use a separate application for the database.

StefanScherer commented on language/golang/develop.md:
````diff
@@ -0,0 +1,475 @@
+Let’s test that our application is connected to the database and is able to save key-value pairs. We'll reuse the test input from the [Build your Go image](build-images.md) section:
+
+```shell
+$ curl --request POST \
+  --url http://localhost:8000/add
+  --header 'content-type: application/json'
+  --data '{"key": "name", "value": "Docker"}'
````
Backslashes are missing.

```suggestion
$ curl --request POST \
  --url http://localhost:8000/add \
  --header 'content-type: application/json' \
  --data '{"key": "name", "value": "Docker"}'
```

olliefr replied:
Yes, how odd that I've missed them... Thank you! Now it's fixed.

StefanScherer commented on language/golang/develop.md:
````diff
@@ -0,0 +1,475 @@
+Now let’s test our API endpoint. Run the following curl command.
+
+```shell
+$ curl --request GET --url http://localhost:8080/
````
The rest-server is using port 8000 now configured in the docker-compose.dev.yml

```suggestion
$ curl --request GET --url http://localhost:8000/
```

StefanScherer reviewed (changes requested):
Hey @olliefr !
Thank you very much for the updates. It's getting even better. 😍 
I walked through the guide until running the tests which fails for me 
```
$ go test -v ./...
=== RUN   TestRespondsWithLove
    main_test.go:47: container not ready, waiting...
    main_test.go:47: container not ready, waiting...
```
probably not finding the database, the container log shows me `pedantic_cray
docker-gs-ping:test
EXITED (1)




2021/05/03 06:38:25 dial tcp 127.0.0.1:5432: connect: connection refused`

But other than that it looks excellent. Thanks for your time and effort. Looking forward to the final updates.

I left a few suggestions where I found minor issues.

olliefr commented:
Nearly there! The guide is _nearly_ complete. What I am going to do next:

* Address the points raised so far in the PR review.
* Read through the guide, in order, and run the commands on a new VM, recording with https://asciinema.org/.
* ???
* Profit! :shipit: 🤣

olliefr commented:
> 5432

This is weird, it's Postgres port... but Postgres is not used in that example. I'll have another look at _all_ examples on a new VM instance, hopefully this will catch any problems which are invisible on my dev machine 😄

olliefr commented:
OK, I'm reasonably happy with it now. All modules are complete. I'd love to run all examples through asciinema on a fresh VM, if I get some free time later this month.

Please feel free to adjust the content as you see fit, I'm happy to assist. 
The main branch diverged quite far and I'm not sure what the policy is on squashing commits, so please advise on how to proceed.

This proved to be way more difficult than I had anticipated 😂 I now think it would have been so much easier to record some screencasts with this content, instead - writing (semi)formal documentation is so hard because consistent style is hard to maintain. But I did my best. I certainly learned _a lot_ in the process!

Let's get it proofread and merged! :shipit: 🚀 Give me a shout any time.

Cheers)

StefanScherer commented:
Hi @olliefr Thank you once again for this excellent contribution!
I finally had the time to walk through the steps again and you created a very good example to walk through all the aspects.
I see only a small conflict in one of the files, but other than that it LGTM to merge it and ship it! 🎉

usha-mandya commented:
Thanks for reviewing this @StefanScherer.

@olliefr If you could fix the conflict, we'll be able to merge the PR. We'd also like to mention your name in the tutorial. I can do a follow-up PR on that once the main set of commits are merged. 

Thanks again for your contribution. We really appreciate it :)

olliefr commented:
Good news, everyone! I've fixed the conflict in the TOC file and re-requested the review. Please let me know if there is anything else. @StefanScherer @usha-mandya

StefanScherer reviewed (approved):
LGTM 🚀 🎉 
Thank you so much!

usha-mandya reviewed (approved):
LGTM. Thank you SO much @olliefr!
